[InstCombine] Fix #163110: Support peeling off matching shifts from icmp operands via canEvaluateShifted

Michael-Chen-NJU opened the pull request:
Consider a pattern like `icmp (shl nsw X, L), (add nsw (shl nsw Y, L), K)`. When the constant K is a multiple of 2^L, this can be simplified to `icmp X, (add nsw Y, K >> L)`. 
This patch extends canEvaluateShifted to support `Instruction::Add` and updates its signature to accept `Instruction::BinaryOps` instead of a boolean. This change allows the function to distinguish between LShr and AShr requirements, ensuring that information is preserved according to the signedness and overflow flags (nsw/nuw) of the operands. 
The logic is integrated into `foldICmpCommutative` to enable peeling off matching shifts from both sides of a comparison even when an offset is present.

Fixes: #163110

Michael-Chen-NJU committed:
[InstCombine] Add test case for fold (X << 5) == ((Y << 5) + 32)

Michael-Chen-NJU committed:
[InstCombine] Optimize icmp with shl and add by folding (X << Log2) == ((Y << Log2) + K) into X == (Y + 1)

dtcxzyw reviewed (commented):
Can we perform this fold with `canEvaluateShifted` + `getShiftedValue`? You can use the known bits to estimate the number of bits to be shifted.

Michael-Chen-NJU committed:
[InstCombine] Add tests for multi-use cases and vector operations in icmp-shl-add optimization

Michael-Chen-NJU committed:
[InstCombine] Refactor icmp folding logic for shl and add operations with APInt

Michael-Chen-NJU committed:
Merge branch 'main' into instcombine/i163110-scaled-equality

Michael-Chen-NJU committed:
Merge branch 'llvm:main' into instcombine/i163110-scaled-equality

Michael-Chen-NJU commented:
> Can we perform this fold with `canEvaluateShifted` + `getShiftedValue`? You can use the known bits to estimate the number of bits to be shifted.

Regarding your suggestion to use canEvaluateShifted and getShiftedValue:

My current approach for folding the pattern` (X << L) == ((Y << L) + K)` into `X == (Y + 1)` is a direct pattern match and substitution in `foldICmpEquality`. Since this optimization relies on the exact structure where $K = 2^L$ and requires the nsw flag for the algebraic simplification (effectively dividing by $2^L$), I thought a targeted pattern match was the most straightforward way to implement this specific, proven optimization.

Are you suggesting we should aim for **a more general, recursive shift distribution** through these APIs?If so, I'm a bit unclear:
Is this specific optimization expected to be handled by a more generic shift distribution rule?Or are you suggesting that using `getShiftedValue` is necessary to handle other, potentially more complex, but unproven patterns that might contain this specific case as a sub-pattern?

If the goal is just to implement the $X = Y + 1$ simplification for this exact pattern, does my current code not already meet the requirement?Any further clarification on the intended scope would be super helpful!

Michael-Chen-NJU committed:
Merge branch 'llvm:main' into instcombine/i163110-scaled-equality

Michael-Chen-NJU committed:
[InstCombine] Fold icmp (shl nsw X, L), (add nsw (shl nsw Y, L), K).

Michael-Chen-NJU commented:
ping

Michael-Chen-NJU committed:
using canEvaluateShifted & getShiftedValue

Michael-Chen-NJU committed:
update tests

Michael-Chen-NJU committed:
Handle Add in getShiftedValue via operand remapping

Michael-Chen-NJU commented:
ping.

Michael-Chen-NJU committed:
[InstCombine]Improved icmp(shl, add(shl, K)) folding by supporting both NSW/AShr and NUW/LShr paths

Michael-Chen-NJU committed:
[InstCombine] Consolidate shift evaluation logic and fix poison flag propagation

Michael-Chen-NJU committed:
Enhance logic in canEvaluateShifted and update test cases

Michael-Chen-NJU committed:
[InstCombine] Refactor shift evaluation logic to use opcode directly and add regression test for fuzzing issue

Michael-Chen-NJU committed:
Refactor canEvaluateShifted to improve handling of shift operations and add regression test for unsigned comparison issue

Michael-Chen-NJU commented:
> Miscompilation reproducer: https://alive2.llvm.org/ce/z/b7sHPP
> 
> ```
> 
> ----------------------------------------
> define i1 @src(i32 %x, i32 %y) {
> #0:
>   %shlx = shl nuw i32 %x, 5
>   %shly = shl nsw i32 %y, 5
>   %add = add nsw i32 %shly, 32
>   %cmp = icmp ugt i32 %shlx, %add
>   ret i1 %cmp
> }
> =>
> define i1 @tgt(i32 %x, i32 %y) {
> #0:
>   %shly = and i32 %y, 134217727
>   %add = add nsw nuw i32 %shly, 1
>   %cmp = icmp ugt i32 %x, %add
>   ret i1 %cmp
> }
> Transformation doesn't verify!
> 
> ERROR: Value mismatch
> 
> Example:
> i32 %x = #x00000001 (1)
> i32 %y = #xffffffff (4294967295, -1)
> 
> Source:
> i32 %shlx = #x00000020 (32)
> i32 %shly = #xffffffe0 (4294967264, -32)
> i32 %add = #x00000000 (0)
> i1 %cmp = #x1 (1)
> 
> Target:
> i32 %shly = #x07ffffff (134217727)
> i32 %add = #x08000000 (134217728)
> i1 %cmp = #x0 (0)
> Source value: #x1 (1)
> Target value: #x0 (0)
> ```

Hi @dtcxzyw,

I have fixed this miscompilation by refactoring `canEvaluateShifted` and added your fuzzer-generated case as a regression test.

Verified with Alive2: https://alive2.llvm.org/ce/z/yqWhxK

Michael-Chen-NJU committed:
Refactor getShiftedValue to drop poison generating flags for left shifts

Michael-Chen-NJU committed:
[InstCombine] Refactor canEvaluateShifted to improve AShr handling

Michael-Chen-NJU commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -677,6 +702,10 @@ static Value *foldShiftedShift(BinaryOperator *InnerShift, unsigned OuterShAmt,
+    if (IsInnerShl &&
+        ((OuterOpc == Instruction::AShr && InnerShift->hasNoSignedWrap()) ||
+         (OuterOpc == Instruction::LShr && InnerShift->hasNoUnsignedWrap())))
+      return InnerShift->getOperand(0);
```
Hi @dtcxzyw, I have updated the code to return nullptr for AShr to ensure it doesn't hit the follow-up masking logic.

But I have a quick question: In foldShiftedShift, I kept the `InnerShift->hasNoSignedWrap()` check for `AShr`. Given that canEvaluateShiftedShift now strictly filters out AShr without nsw, this check is technically redundant in the current call path.

Do you prefer keeping it here for defensive purposes to keep the function's logic self-contained, or should I simplify it since the precondition is already guaranteed by the caller? Thanks!

Michael-Chen-NJU committed:
Refactor foldShiftedShift to ensure correct conditions for shift operations

dtcxzyw reviewed (approved):
LGTM. Thanks!
Please update the PR title and description before landing this patch.

Michael-Chen-NJU committed:
update comment

Michael-Chen-NJU commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
Hi @nikic, I have updated the doc comments for canEvaluateShifted to explicitly clarify the current semantics. Does this clarification address your concerns and is it sufficient to move forward with landing this patch?

Michael-Chen-NJU committed:
update

Michael-Chen-NJU committed:
update lossy and lossless
